sql/join.go

```diff
@@ -0,0 +1,146 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const joinBatchSize = 100
+
+// An indexJoinNode implements joining of results from an index with the rows
+// of a table. The index side of the join is pulled first and the resulting
+// rows are used to lookup rows in the table. The work is batched: we pull
+// joinBatchSize rows from the index and use the primary key to construct spans
+// that are looked up in the table.
+type indexJoinNode struct {
+	index            *scanNode
+	table            *scanNode
+	primaryKeyPrefix proto.Key
+	colIDtoRowIndex  map[ColumnID]int
+	err              error
+}
+
+func makeIndexJoin(indexScan *scanNode) (*indexJoinNode, error) {
+	// Copy the index scan node into a new table scan node and reset the fields
+	// that were set up for the index scan.
+	table := &scanNode{}
+	*table = *indexScan
+	table.index = &table.desc.PrimaryIndex
+	table.spans = nil
+	table.reverse = false
+	table.isSecondaryIndex = false
+	table.initOrdering()
+
+	// Reset the render targets for the index scan to be exactly the primary key
+	// columns.
+	indexScan.render = nil
+	colIDtoRowIndex := map[ColumnID]int{}
+	for i, colName := range table.desc.PrimaryIndex.ColumnNames {
+		sexpr := parser.SelectExpr{
+			Expr: &parser.QualifiedName{Base: parser.Name(colName)},
+		}
+		if err := indexScan.addRender(sexpr); err != nil {
+			return nil, err
+		}
+		colIDtoRowIndex[table.desc.PrimaryIndex.ColumnIDs[i]] = i
+	}
+	indexScan.initOrdering()
+
+	primaryKeyPrefix := proto.Key(MakeIndexKeyPrefix(table.desc.ID, table.index.ID))
+
+	return &indexJoinNode{
+		index:            indexScan,
+		table:            table,
+		primaryKeyPrefix: primaryKeyPrefix,
+		colIDtoRowIndex:  colIDtoRowIndex,
+	}, nil
+}
+
+func (n *indexJoinNode) Columns() []string {
+	return n.table.Columns()
+}
+
+func (n *indexJoinNode) Ordering() []int {
+	return n.table.Ordering()
+}
+
+func (n *indexJoinNode) Values() parser.DTuple {
+	return n.table.Values()
+}
+
+func (n *indexJoinNode) Next() bool {
+	// First, try to pull rows from the table.
+	if n.table.kvs != nil && n.table.Next() {
+		return true
+	}
+	if n.err = n.table.Err(); n.err != nil {
+		return false
+	}
+
+	// The table is out of rows. Pull primary keys from the index.
+	n.table.kvs = nil
+	n.table.kvIndex = 0
+	n.table.spans = n.table.spans[0:0]
+
+	for len(n.table.spans) < joinBatchSize {
```

Comment:
Heh, I originally had it that way, but wanted to make it explicit that the loop existed when we had batched up enough spans.